Remove expansion of empty index arguments in RoutingTable

cbuescher opened the pull request:
RoutingTables activePrimaryShardsGrouped(), allActiveShardsGrouped() and
allAssignedShardsGrouped() methods treated empty index array input
parameters as meaning "all" indices and expanded to the routing maps
keyset. However, the expansion of index names is now already done in
MetaData#concreteIndices(). Returning an empty index name list here
when a wildcard pattern didn't match any index name could lead to
problems like #9081 because the RoutingTable still expanded this
list of names to "_all". In case of e.g. the recovery endpoint this
could lead to problems.

This fix removes the index name expansion in RoutingTable and introduces
some more checks for preventing NPEs in MetaData#concreteIndices().

Closes #9081

javanna commented on rest-api-spec/test/indices.recovery/10_basic.yaml:
```diff
@@ -1,9 +1,6 @@
 
-  - skip:
-      features: gtelte
-
```
if you rebase this change should go away right?

cbuescher replied:
Yes, I deleted that on master already. Should also maybe delete the gtelte on 1.x branch?

javanna replied:
yes I think so, should be done on 1.x too

cbuescher replied:
done

javanna commented on src/main/java/org/elasticsearch/cluster/metadata/MetaData.java:
```diff
@@ -675,6 +675,14 @@ public String resolveIndexRouting(@Nullable String routing, String aliasOrIndex)
+            }
+            else return Strings.EMPTY_ARRAY;
+        }
+
```
can you explain what you are trying to achieve with this code block? Does it actually change the result of calling `concreteIndices`? Seems like it doesn't, il wildcards don't need to be expanded _all is treated as a concrete index, but it's not there, then we look at allowNoIndices anyway to decide whether we want to throw exception or just return empty. What am I missing?

cbuescher replied:
When both indicesOptions.expandWildcardsOpen() and indicesOptions.expandWildcardsClosed() are both false for some reason, concrete indices falls through to here. If the 'aliasOrIndices' is an '_all' pattern we want to return here. Otherwise the existing implementation runs into a Null Pointer a few lines below

```
if (aliasesOrIndices.length == 1)
```
At least that's what happens in the randomized test for this method when I introduced this. It's an edge case and the nesting of if-statements here is not nice, but there are so many different cases depending on the indicesOptions settings.

javanna replied:
ah I see, that's because you pass in `null` as second argument. I guess in practice most of the apis fall into the expand wildcards case, that is why we have never seen this problem in reality. 

I think you should just do:

```
if (aliasesOrIndices == null || aliasesOrIndices.length == 0)
```
then instead of
```
if (isAllIndices(aliasesOrIndices)) 
```

and adapt the error message. This doesn't have to do with the `_all` expression, which is treated as a special wildcard.

Note that we were previously missing to check `allowNoIndices` when `aliasesOrIndices` is empty... good test and good change! would be nice to figure out what the impact of this is, I think it might deserve the breaking label.... actually, maybe I would even pull out this change and get it in separately. What do you think?

javanna replied:
ping @cbuescher I think you haven't seen this comment ;)

javanna commented on src/main/java/org/elasticsearch/cluster/metadata/MetaData.java:
```diff
@@ -675,6 +675,14 @@ public String resolveIndexRouting(@Nullable String routing, String aliasOrIndex)
+            if (!indicesOptions.allowNoIndices()) {
+                throw new IndexMissingException(new Index("_all"));
+            }
+            else return Strings.EMPTY_ARRAY;
```
regardless of the below comment, use brackets here please and move the `else` the to the previous line.

javanna commented on src/main/java/org/elasticsearch/cluster/routing/RoutingTable.java:
```diff
@@ -228,15 +203,12 @@ public GroupShardsIterator allAssignedShardsGrouped(String[] indices, boolean in
                 continue;
                 // we simply ignore indices that don't exists (make sense for operations that use it currently)
-//                throw new IndexMissingException(new Index(index));
+                // throw new IndexMissingException(new Index(index));
```
what happened here? we just want to leave this commented out? maybe we want to even remove this line then?

cbuescher replied:
this seems to be my code formatter at work here

javanna replied:
I would leave the first comment line but remove the commented out throw exception

javanna commented on src/test/java/org/elasticsearch/action/IndicesRequestTests.java:
```diff
@@ -432,6 +439,24 @@ public void testRefresh() {
+        response = internalCluster().clientNodeClient().admin().indices().refresh(refreshRequest).actionGet();
+
+        clearInterceptedActions();
+        assertThat("Empty index arrays should match any index, so all shards should be returned", response.getTotalShards(), equalTo(this.totalNumShards));
```
wondering if these additional checks really belong here. IndicesRequestTests is a different beast, that tests indices resolution. I would probably move to a new test class that is more about testing responses and number of shards.

cbuescher replied:
I'll go and try to use this as yaml tests then

javanna commented on src/test/java/org/elasticsearch/action/IndicesRequestTests.java:
```diff
@@ -456,6 +481,12 @@ public void testRecovery() {
+        RecoveryResponse response = internalCluster().clientNodeClient().admin().indices().recoveries(recoveryRequest).actionGet();
+
+        clearInterceptedActions();
+        assertThat("'v*' should not match any indices, so no shard should be returned", response.getTotalShards(), equalTo(0));
```
same as above

javanna commented on src/test/java/org/elasticsearch/action/IndicesRequestTests.java:
```diff
@@ -480,6 +511,12 @@ public void testIndicesStats() {
+        IndicesStatsResponse response = internalCluster().clientNodeClient().admin().indices().stats(indicesStatsRequest).actionGet();
+
+        clearInterceptedActions();
+        assertThat("'v*' should not match any indices, so no shard should be returned", response.getTotalShards(), equalTo(0));
```
same as above, also I wonder if we could unit test `RoutingTable` instead of needing an integration tests to verify how many shards are touched...

javanna commented on src/test/java/org/elasticsearch/cluster/metadata/MetaDataTests.java:
```diff
@@ -579,76 +592,145 @@ public void concreteIndicesWildcardExpansion() {
+            } catch (IndexMissingException e) {
+                if (indicesOptions.allowNoIndices())
+                    fail("calling concreteIndices for all indices with allowNoIndices option should not throw exception");
+            }
```
I think it is not clear here exactly when IndexMissingException is expected to be thrown or not. I would rather move the if on top and have different asserts path based on that. FOr the expected exception one you can then do:

```
try {
  //do something
  fail("shouldn't get here");
} catch (IndexMissingException e) {
  //assert on exception
}

```

javanna commented on src/test/java/org/elasticsearch/cluster/metadata/MetaDataTests.java:
```diff
@@ -579,76 +592,145 @@ public void concreteIndicesWildcardExpansion() {
+            } catch (IndexMissingException e) {
+                if (indicesOptions.expandWildcardsOpen() || indicesOptions.expandWildcardsClosed() ||
+                        indicesOptions.allowNoIndices())
+                    fail("calling concreteIndices with non matchon wildcard and allowNoIndices=true should not throw exception");
```
same as above

javanna commented on src/test/java/org/elasticsearch/cluster/metadata/MetaDataTests.java:
```diff
@@ -579,76 +592,145 @@ public void concreteIndicesWildcardExpansion() {
+                if (indicesOptions.allowNoIndices())
+                    fail("not expecting exeption when either expandWildcardsOpen, expandWildcardsClose or "
+                            + "allowNoIndices is true");
+            }
```
same as above

javanna commented on src/main/java/org/elasticsearch/cluster/routing/RoutingTable.java:
```diff
@@ -117,61 +117,39 @@ public RoutingTableValidation validate(MetaData metaData) {
-        ArrayList<ShardIterator> set = new ArrayList<>();
-        if (indices == null || indices.length == 0) {
-            indices = indicesRouting.keySet().toArray(new String[indicesRouting.keySet().size()]);
+    public List<ShardRouting> allShards(String index) throws IndexMissingException {
```
why move from GroupShardsIterator to List<ShardRouting> ?

cbuescher replied:
The diff is missleading here I think. I split `allShards(String... indices)` into `allShards()` and `List<ShardRouting> allShards(String index)`. My idea was to get rid of the ambigous (String... indices) signature to have one method where it is clear that the caller really wants _all indices and one where he can get one single index, and then have the first call the second in a loop (basically splitting the existing code from `allShards(String... indices)`.

I removed `GroupShardsIterator allShardsGrouped(String... indices)` completely because I didn't find any caller in the core ES codebase. The diff just mixed the changes I think, so it is a bit hard to read.

javanna replied:
I see! I got the split between the two methods, I like. I had missed the removal of `allShardsGrouped`, good!

javanna commented:
I like the change! Left a bunch of comments

cbuescher commented:
I adressed your comments, will revert the changes in IndicesRequestTests and create rest tests for that later. Let me know if my other replies make sense.

cbuescher commented:
@javanna pushed a couple of changes regarding your comments. For one, I reverted the changes to IndicesRequestTests and introduced rest tests for these cases. Other changes address your comments to code style mostly. Let me know what you think.

javanna commented on src/test/java/org/elasticsearch/cluster/metadata/MetaDataTests.java:
```diff
@@ -579,76 +592,155 @@ public void concreteIndicesWildcardExpansion() {
+     * test resolving _all pattern (null, empty arry or "_all") for random IndicesOptions
+     */
+    @Test
+    public void concreteIndicesAllpatternRandom() {
```
concreteIndicesAllpatternRandom=>concreteIndicesAllPatternRandom

javanna commented:
Hi @cbuescher I did another review round. it's close.

I am not sure if we should rely solely on REST tests to test the `RoutingTable` changes. REST tests are great but I would love to have java unit tests around them. The tests that you previously added to `IndicesRequestTests` were not bad, just in the wrong place. Also, I wondered if we could move them to a more isolated unit tests instead of having to send requests and read responses to find out how many shards were touched. Replacing with REST tests makes tests even less isolated, makes sense?

cbuescher commented:
@javanna I adressed two of your comments and started adding a RoutingTable unit test, but it's still a very rudimentary test. One thing I'm currently not sure about is how to set up the RoutingTable for the test so that the shard routings that are created are in a different state than "unassigned". When I figured that out I will extend the test. Any suggestion on how to do that in a simple unit test way are.

javanna commented on src/test/java/org/elasticsearch/cluster/metadata/MetaDataTests.java:
```diff
@@ -579,76 +592,155 @@ public void concreteIndicesWildcardExpansion() {
+                assertThat(concreteIndices, notNullValue());
+                int expectedNumberOfIndices = 0;
+                if (indicesOptions.expandWildcardsOpen())
+                    expectedNumberOfIndices += 2;
```
brackets please

javanna commented on src/test/java/org/elasticsearch/cluster/metadata/MetaDataTests.java:
```diff
@@ -579,76 +592,155 @@ public void concreteIndicesWildcardExpansion() {
+                if (indicesOptions.expandWildcardsOpen())
+                    expectedNumberOfIndices += 2;
+                if (indicesOptions.expandWildcardsClosed())
+                    expectedNumberOfIndices += 1;
```
here too

javanna commented on src/test/java/org/elasticsearch/cluster/metadata/MetaDataTests.java:
```diff
@@ -579,76 +592,155 @@ public void concreteIndicesWildcardExpansion() {
+     */
+    @Test
+    public void concreteIndicesWildcardEmptyRandom() {
+        for (int i = 0; i < 100; i++) {
```
wondering if we really need to repeat this 100 times. Seems a little too high, wouldn't 10 be enough? Or even just rely on the fact that tests run continously. An alternative would be to use the `@Repeat` annotation instead of the loop. What do you think?

cbuescher replied:
Sure, will decrease that

javanna commented on src/test/java/org/elasticsearch/cluster/routing/RoutingTableTest.java:
```diff
@@ -0,0 +1,112 @@
+
+import static org.hamcrest.Matchers.is;
+
+public class RoutingTableTest extends ElasticsearchAllocationTestCase  {
```
does this really need to extend `ElasticsearchAllocationTestCase`? I think `ElasticsearchTestCase` should be enough.

cbuescher replied:
I think I was trying to use this to be able to change the ShardRoutingState in the test setup to be able to test other than just "unassigned" shards. I haven't yet figured out how to do that though.

javanna commented:
Thanks for the update @cbuescher I had another look, looks good. You are on the right track with the `RoutingTable` test, looks great.

cbuescher commented:
Added changes according to latest review comments and extended RoutingTableTest so it covers moving through different ShardRoutingStates now.

javanna commented on src/test/java/org/elasticsearch/cluster/metadata/MetaDataTests.java:
```diff
@@ -579,76 +592,157 @@ public void concreteIndicesWildcardExpansion() {
     public void testIsExplicitAllIndices_wildcard() throws Exception {
-        MetaData metaData = MetaData.builder().build();
-        assertThat(metaData.isExplicitAllPattern(new String[]{"*"}), equalTo(false));
+        assertThat(MetaData.isExplicitAllPattern(new String[]{"*"}), equalTo(false));
```
if you get merge conflicts on these MetaData calls that should have been static, don't worry, I think I recently merged a PR that fixed just that.

javanna commented on src/test/java/org/elasticsearch/cluster/routing/RoutingTableTest.java:
```diff
@@ -0,0 +1,248 @@
+        try {
+            assertThat(this.testRoutingTable.allShards("not_existing").size(), is(0));
+            fail("Exception expected when calling allShards() with non existing index name");
+        } catch (IndexMissingException e) {
```
if the exception is the last thing that happens you may replace it with @Test(expected = IndexMissingException.class)

javanna commented on src/test/java/org/elasticsearch/cluster/routing/RoutingTableTest.java:
```diff
@@ -0,0 +1,248 @@
+        try {
+            this.testRoutingTable.activePrimaryShardsGrouped(new String[] {TEST_INDEX_1, "not_exists"}, true);
+            fail("Calling with non-existing index name should raise IndexMissingException");
+        } catch (IndexMissingException e) {
```
same as above

javanna replied:
scratch that, maybe it makes sense to keep it as-is given that the text is quite long and we want to make sure that the exception happens there, and not before that line

javanna commented on src/test/java/org/elasticsearch/cluster/metadata/MetaDataTests.java:
```diff
@@ -579,76 +592,157 @@ public void concreteIndicesWildcardExpansion() {
+     * test resolving wildcard pattern that matches no index of alias for random IndicesOptions
+     */
+    @Test
+    public void concreteIndicesWildcardEmptyRandom() {
```
can you prepend the method name with `test`, just to make sure ?

cbuescher replied:
I'll change that for all the tests (also the existing ones) in MetaDataTests then.

javanna commented on src/test/java/org/elasticsearch/cluster/metadata/MetaDataTests.java:
```diff
@@ -579,76 +592,157 @@ public void concreteIndicesWildcardExpansion() {
+     * test resolving _all pattern (null, empty arry or "_all") for random IndicesOptions
+     */
+    @Test
+    public void concreteIndicesAllPatternRandom() {
```
can you prepend the method name with test, just to make sure that it gets run ?

javanna commented on src/test/java/org/elasticsearch/cluster/metadata/MetaDataTests.java:
```diff
@@ -579,76 +592,157 @@ public void concreteIndicesWildcardExpansion() {
+                try {
+                    metadata.concreteIndices(indicesOptions, allIndices);
+                    fail("calling concreteIndices for _all indices with allowNoIndices=false should throw exception");
+                } catch (ElasticsearchIllegalArgumentException | IndexMissingException e) {
```
when should it throw EIAE and when index missing?

cbuescher replied:
I chose to throw the ElasticsearchIllegalArgumentException when the wildcard-expansion options are all off (expandWildcardsOpen, expandWildcardsClosed), allowNoIndices is also off and the method call is made with empty or null indices-names list. The IndexMissingException is thrown when wildcard expansion happend but returned nothing (and allowNoIndices is off).

Should I change this to IndexMissingException in the first case also? Feels to me more like a mistake by the callern when he requests "_all" by way of null or empty list but doesn't turn wildcard expansion on.

javanna replied:
I see, I commented above on the MetaData change. As for this test, I would prefer to better define what is expected when. for instanche two different catch blocks depending on the indices options and what is expected.

javanna commented:
Had another look, left a bunch of super minor comments, this is very close.

cbuescher commented:
I left the two exception types in MetaData.concreteIndices() but extended the test to check for throwing the correcty type. Can also make both IndexMissingExceptions, but that one wants the name of the missing index and it realy is more like a call to the function with mismatch in arguments. I hope its okay to give the IndicesOptions a simple toString() to see whats going on in case of an exception.

javanna commented on src/main/java/org/elasticsearch/cluster/metadata/MetaData.java:
```diff
@@ -675,6 +675,16 @@ public String resolveIndexRouting(@Nullable String routing, String aliasOrIndex)
+
+        if (aliasesOrIndices == null || aliasesOrIndices.length == 0) {
+            if (!indicesOptions.allowNoIndices()) {
+                throw new ElasticsearchIllegalArgumentException("Null or zero length argument for list of index or alias names "
```
we throw `IndexMissingException` all the time when `allowNoIndices` is set to true. I don't think we should act differently here.

javanna replied:
but now that I read your comment below, I see why it makes sense. Fine with illegal argument, I would only change the message to something like "no indices were specified but wildcard expansion is disabled" , shorter and more user oriented, what do you think?

javanna commented on src/main/java/org/elasticsearch/action/support/IndicesOptions.java:
```diff
@@ -245,6 +245,11 @@ private static byte toByte(boolean ignoreUnavailable, boolean allowNoIndices, bo
+    @Override
+    public String toString() {
+        return "id: " + this.id;
+    }
```
I don't think this is needed, also not meaningful as it would print out a "random" number, it's not something that users can understand

cbuescher commented:
We agreed on pulling the changes in MetaData and MetaDataTests out into a separate PR. Will open that soon. This PR now should only remove the expansion of empty or null index name lists in RoutingTable.

cbuescher commented:
@javanna With the removal of the changes in MetaData to #10339, anything else to change here?

javanna commented:
LGTM @cbuescher thanks
